src/main/java/edu/iu/terracotta/service/app/impl/AssessmentServiceImpl.java

```diff
@@ -430,4 +460,114 @@ public AssessmentDto duplicateAssessment(long assessmentId, long treatmentId) th
+        }
+
+        return assessment;
+    }
+
+    @Override
+    public Assessment getAssessmentByGroupId(Long experimentId, String canvasAssignmentId, Long groupId) throws AssessmentNotMatchingException {
+        Assignment assignment = allRepositories.assignmentRepository.findByExposure_Experiment_ExperimentIdAndLmsAssignmentId(experimentId, canvasAssignmentId);
+
+        if (assignment == null){
+            throw new AssessmentNotMatchingException("Error 127: This assignment does not exist in Terracotta for this experiment");
+        }
+
+        Optional<ExposureGroupCondition> exposureGroupCondition = allRepositories.exposureGroupConditionRepository.getByGroup_GroupIdAndExposure_ExposureId(groupId, assignment.getExposure().getExposureId());
+
+        if (!exposureGroupCondition.isPresent()){
+            throw new AssessmentNotMatchingException("Error 130: This assignment does not have a condition assigned for the participant group.");
+        }
+
+        return retrieveTreatmentAssessment(exposureGroupCondition.get().getCondition().getConditionId(), assignment.getAssignmentId());
+    }
+
+    @Override
+    public Assessment getAssessmentByConditionId(Long experimentId, String canvasAssignmentId, Long conditionId) throws AssessmentNotMatchingException {
+        Assignment assignment = allRepositories.assignmentRepository.findByExposure_Experiment_ExperimentIdAndLmsAssignmentId(experimentId, canvasAssignmentId);
+
+        if (assignment == null) {
+            throw new AssessmentNotMatchingException("Error 127: This assignment does not exist in Terracotta for this experiment");
+        }
+
+        return retrieveTreatmentAssessment(conditionId, assignment.getAssignmentId());
+    }
+
+    private Assessment retrieveTreatmentAssessment(long conditionId, long assignmentId) throws AssessmentNotMatchingException {
+        List<Treatment> treatments = allRepositories.treatmentRepository.findByCondition_ConditionIdAndAssignment_AssignmentId(conditionId, assignmentId);
+
+        if (treatments.isEmpty()) {
+            throw new AssessmentNotMatchingException("Error 131: This assignment does not have a treatment assigned.");
+        }
+
+        if (treatments.size() > 1) {
+            throw new AssessmentNotMatchingException("Error 132: This assignment has ambiguous treatments. Please contact a Terracotta administrator");
+        }
+
+        if (treatments.get(0).getAssessment() == null) {
+            throw new AssessmentNotMatchingException("Error 133: The treatment for this assignment does not have an assessment created");
+        }
+
+        return treatments.get(0).getAssessment();
+    }
+
+    @Override
+    public AssessmentDto viewAssessment(long experimentId, SecuredInfo securedInfo) throws ExperimentNotMatchingException, ParticipantNotMatchingException, AssessmentNotMatchingException {
+        Optional<Experiment> experiment = allRepositories.experimentRepository.findById(experimentId);
```

Comment:
This method needs some additional functionality that is currently present in `AssignmentServiceImpl.launchAssignment`. `launchAssignment` was basically the entry point of a student into a Terracotta experiment. This is where a student consent is checked and the student is assigned to a group. And the group determines what treatment/assessment the student will get. But now, with this redesign, the `viewAssignment` method becomes the entry point. 

So essentially all of the logic in `launchAssignment` that comes before getting the assessment needs to be here as well. Specifically:

- call refreshParticipants. This student may have joined the course recently so they may not yet have a participant record.
- update participant's consent. https://github.com/terracotta-education/terracotta/blob/development/src/main/java/edu/iu/terracotta/service/app/impl/AssignmentServiceImpl.java#L440-L449
- is the student has consented, assign the participant to a group.
- save the participant record

Reply:
Done!